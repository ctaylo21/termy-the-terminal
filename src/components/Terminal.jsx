import React, { Component } from 'react';
import PropTypes from 'prop-types';
import History from './History';
import Input from './Input';

export default class Terminal extends Component {
  constructor(props) {
    super(props);
    this.currentCommandId = 0;
    this.state = {
      history: [],
      inputValue: '',
    };
  }

  handleKeyPress = (event) => {
    if (event.key === 'Enter') {
      const { history } = this.state;
      const updatedHistory = history.concat({
        id: this.currentCommandId++,
        value: event.target.value,
      });

      this.setState({
        history: updatedHistory,
        inputValue: '',
      });
    }
  }

  render() {
    const { history, inputValue } = this.state;
    const { inputPrompt } = this.props;
    return (
      <>
        <History history={history} />
        <Input inputPrompt={inputPrompt} handleKeyPress={this.handleKeyPress} inputValue={inputValue} />
      </>
    );
  }
}

Terminal.propTypes = {
  inputPrompt: PropTypes.string.isRequired,
};
