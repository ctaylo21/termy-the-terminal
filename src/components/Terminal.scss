@import '~reset-css';
@import '../styles/vars';

#terminal-wrapper {
  background-color: map-get($themeDark, background);
  box-sizing: content-box;
  color: map-get($themeDark, fontColor);
  font-family: $fontFamily;
  font-size: $fontSize;
  height: 100%;
  padding: 0 1em;

  #history-container {
    li {
      margin-bottom: 1em;
    }
  }

  #input-container {
    padding-top: 1em;

    form {
      display: flex;

      input {
        background: none;
        border: none;
        color: map-get($themeDark, fontColor);
        font-family: $fontFamily;
        font-size: $fontSize;
        outline: none;
        -webkit-appearance: none;

        flex-grow: 100;
        margin: 0 10px;
      }

      #inputPromptChar {
        color: map-get($themeDark, green);
      }
    }
  }
}
