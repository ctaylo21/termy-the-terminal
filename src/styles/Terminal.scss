@import './vars';

#terminal-wrapper {
  background-color: map-get($themeDark, background);
  box-sizing: content-box;
  color: map-get($themeDark, fontColor);
  font-family: map-get($themeDark, fontFamily);
  font-size: map-get($themeDark, fontSize);
  height: 100%;
  overflow: auto;
  padding: 0 1em;

  #input-container {
    padding: 1em 0;

    form {
      display: flex;

      input {
        background: none;
        border: none;
        color: map-get($themeDark, fontColor);
        font-family: map-get($themeDark, fontFamily);
        font-size: map-get($themeDark, fontSize);
        outline: none;
        -webkit-appearance: none;

        flex-grow: 100;
        margin: 0 10px;
      }

      #inputPrompt {
        color: map-get($themeDark, green);
      }
    }
  }

  #history-container {
    & > ul > li > #input-container {
      padding-bottom: 0;
    }

    img {
      margin-top: 10px;
    }

    #help-container ul > li {
      margin: 5px 0;

      & > span.help-command-name {
        color: map-get($themeDark, purple);
      }
    }

    ul {
      list-style-type: none;
      margin: 0;
      padding: 0;
    }

    ul.terminal-ls-list {
      margin: 10px 0 0 10px;

      li {
        display: flex;
        align-items: center;

        svg {
          margin-right: 20px;
        }

        &.ls-folder {
          color: map-get($themeDark, orange);

          svg {
            fill: map-get($themeDark, orange);
            height: 24px;
            width: 24px;
          }
        }

        &.ls-file {
          color: map-get($themeDark, blue);

          svg {
            fill: map-get($themeDark, blue);
            margin-top: 2px;
            margin-bottom: 2px;
            height: 20px;
            width: 24px;
          }
        }
      }
    }
  }

  .preview-list {
    span {
      margin: 0px 10px;

      &.auto-preview-folder {
        color: map-get($themeDark, orange);
      }

      &.auto-preview-file {
        color: map-get($themeDark, blue);
      }

      &.active {
        background-color: map-get($themeDark, highlight);
      }
    }


  }
}
