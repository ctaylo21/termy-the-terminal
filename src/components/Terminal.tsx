import React, { ChangeEvent, Component, FormEvent } from 'react';
import { History } from './History';
import Input from './Input';
import { cd, ls } from '../services';
import './Terminal.scss';

export class Terminal extends Component<TerminalProps, TerminalState> {
  public readonly state: TerminalState = {
    currentCommandId: 0,
    currentPath: '/',
    history: [],
    inputValue: '',
    promptChar: 'â†’',
  };

  private handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      inputValue: event.target.value,
    });
  };

  private handleSubmit = async (
    event: FormEvent<HTMLFormElement>,
  ): Promise<void> => {
    event.preventDefault();

    const commandArgs = this.state.inputValue.split(' ');

    const { history, inputValue } = this.state;

    let result = '';
    switch (commandArgs[0]) {
      case 'cd':
        try {
          const newPath = await cd(
            this.props.fileSystem,
            this.state.currentPath,
            commandArgs[1],
          );

          result = 'cd success';
          this.setState({
            currentPath: newPath,
          });
        } catch (cdException) {
          result = cdException;
        }
        break;
      case 'pwd':
        result = this.state.currentPath;
        break;
      case 'ls':
        const pathToLs = commandArgs[1]
          ? commandArgs[1].startsWith('/')
            ? commandArgs[1]
            : `${this.state.currentPath}/${commandArgs[1]}`
          : this.state.currentPath;

        try {
          const test = await ls(this.props.fileSystem, pathToLs);
          result = JSON.stringify(test);
        } catch (e) {
          result = e;
        }
        break;
      default:
        result = 'Invalid command';
        break;
    }

    const updatedHistory = history.concat({
      id: this.state.currentCommandId,
      result,
      value: inputValue,
    });

    this.setState({
      currentCommandId: this.state.currentCommandId + 1,
      history: updatedHistory,
      inputValue: '',
    });
  };

  public render(): JSX.Element {
    const { currentPath, history, inputValue, promptChar } = this.state;
    return (
      <>
        <History history={history} />
        <Input
          currentPath={currentPath}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          inputValue={inputValue}
          promptChar={promptChar}
        />
      </>
    );
  }
}
