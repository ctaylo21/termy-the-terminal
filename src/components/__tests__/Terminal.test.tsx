import React from 'react';
import { cleanup, fireEvent, render } from 'react-testing-library';
import Terminal from '../Terminal';

afterEach(cleanup);

test('invalid command', () => {
  const { getByLabelText } = render(<Terminal />);
  const input = getByLabelText('terminal-input');

  fireEvent.change(input, { target: { value: 'invalid-command' } });
  fireEvent.submit(input);

  const history = getByLabelText('terminal-history');

  expect(history.innerHTML).toMatchSnapshot();
});
