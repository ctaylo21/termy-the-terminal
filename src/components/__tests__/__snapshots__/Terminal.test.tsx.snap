// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid command 1`] = `"<li>invalid-command : Invalid command</li>"`;
